core/resource/store/store.go

```diff
@@ -47,11 +50,57 @@ func NewFingerprint(f hash.Fingerprint) Fingerprint {
 	return Fingerprint{f}
 }
 
-// UUID is the UUID of the stored blob in the database, this can
+// ID is the ID of the stored blob in the database, this can
 // be used for adding referential integrity from the resource to the stored
 // blob. This can be an object store metadata UUID or a container image metadata
-// storage key.
-type UUID string
+// storage key. It is only one, never both.
+type ID struct {
+	objectStoreUUID               objectstore.UUID
```

Comment:
Should we have a kind, so we can say it's one or the other, rather than checking if a value is empty. Otherwise it becomes very passive.

Reply:
This is basically a union type. A suggested implementation is defined here:

https://www.jerf.org/iri/post/2917/

Could look goofy, but it is indeed elegant: No need to define yet another enum.

```go
type ID interface {
   isStoreID()
}

// Implements empty isStoreID() for whatever id we have. 
// We would need to create an alias for the string type of containerImageMetadataStore.
type containerImageMetadataStoreID string

// To be able to findout which type it is:

func someFuncRelyingOnTYpe(myID ID){
   switch id := miID.(type) {
       case objectstore.UUID: 
            // id is an objectstore.UUID
       case containerImageMetadataStoreID:
            // id is an containerImageMetadataStoreID
   }
}

```

Reply:
I used the interface way quite a bit in SQLair, its ok, but I'm not a huge fan of it, it feels like a misuse of the language construct, even if the type seen externally is a bit nicer. Theres just not really a good way to do it. I'll go with kinds just because its easier to implement.